Readme.md
Inofficial Fork of weeaar_googlesitemap
=======================================

based on official release 0.4.3 from TER:
http://typo3.org/extensions/repository/view/weeaar_googlesitemap

New Features
------------

  * tt_news single page views can now be optionally kept on `sitemap.xml` output (e.g. for dual-mode pages which contain both a `LIST` and `SINGLE` view on the same page and toggle the active tt_news mode based upon parameters passed on URL). Can be used by setting new parameter `tt_news.single_page.*.keepListedAsPage = 1`. See changeset `cbe4a3fe159b09912fe777ec7ad618ce8a9be918` for details.
  * tt_news entries can optionally be filtered by archived/non-archived state. If filtering is desired, set `archive_mode = archived` or `archive_mode = non-archived` on a single view configuration
  * news schema is no longer active by default as it is intended for actual news like magazine articles or press releases which if not the case for every tt_news article
    * enabled via `news_schema = 1` on `single_page` configuration
    * configuration allows setting all tags as described by [Google](https://support.google.com/news/publisher/answer/74288):
      * `news_schema.max_age_days` (optional) - Google requests to list only news which have an age of less than 2 days; you may override that limit using this setting
        * Note: News will still be listed as regular pages after if the maximum age has been exceeded, they just won't get the additional news schema.
      * `news_schema.publication.name` (required) - the publication's name (magazine name, news publisher name, organization, *not* the author)
      * `news_schema.publication.language` (required) - the language used by the publication as default
      * `news_schema.publication.language.<sys_language_uid> = <language name>` (optional) - mapping individual `sys_language_uid`s to language names (BE record UID, not FE `L`-parameter!)
      * `news_schema.keywords_default` (optional) - keywords to use if news article doesn't set any (may be overriden by saving spaces to news); keywords should be [Google news topics](https://support.google.com/news/publisher/answer/116037)
      * `news_schema.keywords_override` (optional) - keywords to use regardless of default or news values; keywords should be [Google news topics](https://support.google.com/news/publisher/answer/116037)
      * `news_schema.access` (optional) - subscription/registration required? assumed free if missing; see [Google help](https://support.google.com/news/publisher/answer/74288) for details
      * `news_schema.genres` (optional) - type of content; see [Google help](https://support.google.com/news/publisher/answer/93992) for details
      * `news_schema.stock_tickers` (optional) - related stock exchanges & ticker IDs; see [Google help](https://support.google.com/news/publisher/answer/74288) for details
    * example:
```
single_page.1 {
    news_schema = 1
    news_schema {
        publication {
            name = Example Magazine
            language = en
            language {
                0 = en
                1 = de
            }
        }
        access = Registration
        genres = UserGenerated
        keywords_default = Health
        #keywords_override = Health
        #stock_tickers = TEST:A
        max_age_days = 5
    }
}
```

Fixes
-----

  * fix for security bulletin [TYPO3-EXT-SA-2014-010](http://typo3.org/teams/security/security-bulletins/typo3-extensions/typo3-ext-sa-2014-010/); credits to initiator of fork [cab services ag](https://github.com/cabservicesag/weeaar_googlesitemap)
  * processing of translated tt_news entries
  * news schema now contains all required tags; see new features above
  * runs under TYPO3 6.2


ChangeLog
2015-11-05 Daniel Neugebauer <dneuge@glutrot.de>
	* new parameter tt_news.single_page.*.recursive to allow getting news recursively
	* news now honor languageParamIf0 setting
	* deleted page translations are no longer shown

2015-03-23 Daniel Neugebauer <dneuge@glutrot.de>
	* compatibility with TYPO3 6.2

2014-11-19 Daniel Neugebauer <dneuge@glutrot.de>
	* check for new parameter tt_news.single_page.*.keepListedAsPage to keep single view pages listed

2014-09-04 Jonas Felix <jf@cabag.ch>
	* version 0.4.4
	* Security fix typo3-ext-sa-2014-010
	* Minor cleanups
	* fixes potential sql injection in backend

07-04-30 Marcel Alburg  <alb@weeaar.com>
	* Bugfix for single language projects (Thanks to Martin Scholl [ms@globalinfinity.de])

06-10-23 Marcel Alburg  <alb@weeaar.com>

	* Initial release


mod1/index.php
<?php

/* * *************************************************************
 *  Copyright notice
 *
 *  (c) 2006 Marcel Alburg <alb@weeaar.com>
 *  All rights reserved
 *
 *  This script is part of the TYPO3 project. The TYPO3 project is
 *  free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  The GNU General Public License can be found at
 *  http://www.gnu.org/copyleft/gpl.html.
 *
 *  This script is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  This copyright notice MUST APPEAR in all copies of the script!
 * ************************************************************* */


// DEFAULT initialization of a module [BEGIN]
unset($MCONF);
require_once('conf.php');
require_once($BACK_PATH . 'init.php');
if (!class_exists('template')) {
	require_once($BACK_PATH . 'template.php');
}

$LANG->includeLLFile('EXT:weeaar_googlesitemap/mod1/locallang.xml');
if (!class_exists('t3lib_SCbase')) {
	require_once(PATH_t3lib . 'class.t3lib_scbase.php');
}
$BE_USER->modAccess($MCONF, 1); // This checks permissions and exits if the users has no permission for entry.
// DEFAULT initialization of a module [END]

/**
 * Module 'GoogleSitemap' for the 'weeaar_googlesitemap' extension.
 *
 * @author	Marcel Alburg <alb@weeaar.com>
 * @package	TYPO3
 * @subpackage	tx_weeaargooglesitemap
 */
class tx_weeaargooglesitemap_module1 extends t3lib_SCbase {

	var $pageinfo;
	var $data = Array();

	/**
	 * Initializes the Module
	 * @return	void
	 */
	function init() {
		global $BE_USER, $LANG, $BACK_PATH, $TCA_DESCR, $TCA, $CLIENT, $TYPO3_CONF_VARS;

		parent::init();
	}

	/**
	 * Adds items to the ->MOD_MENU array. Used for the function menu selector.
	 *
	 * @return	void
	 */
	function menuConfig() {
		global $LANG;
		$this->MOD_MENU = Array(
			'function' => Array(
				'1' => $LANG->getLL('siteconfiguration'),
			)
		);
		parent::menuConfig();
	}

	/**
	 * Main function of the module. Write the content to $this->content
	 * If you chose "web" as main module, you will need to consider the $this->id parameter which will contain the uid-number of the page clicked in the page tree
	 *
	 * @return	[type]		...
	 */
	function main() {
		global $BE_USER, $LANG, $BACK_PATH, $TCA_DESCR, $TCA, $CLIENT, $TYPO3_CONF_VARS;

		// Access check!
		// The page will show only if there is a valid page and if this page may be viewed by the user
		$this->pageinfo = t3lib_BEfunc::readPageAccess($this->id, $this->perms_clause);
		$access = is_array($this->pageinfo) ? 1 : 0;

		if (($this->id && $access) || ($BE_USER->user['admin'] && !$this->id)) {

			// Draw the header.
			$this->doc = t3lib_div::makeInstance('mediumDoc');
			$this->doc->backPath = $BACK_PATH;
			$this->doc->form = '<form action="" method="POST">';

			// JavaScript
			$this->doc->JScode = '
							<script language="javascript" type="text/javascript">
								script_ended = 0;
								function jumpToUrl(URL)	{
									document.location = URL;
								}
							</script>
						';
			$this->doc->postCode = '
							<script language="javascript" type="text/javascript">
								script_ended = 1;
								if (top.fsMod) top.fsMod.recentIds["web"] = 0;
							</script>
						';
			
			$pathCropped = method_exists('TYPO3\CMS\Core\Utility\GeneralUtility', 'fixed_lgd_cs') ? TYPO3\CMS\Core\Utility\GeneralUtility::fixed_lgd_cs($this->pageinfo['_thePath'], 50) : t3lib_div::fixed_lgd_pre($this->pageinfo['_thePath'], 50);
			$headerSection = $this->doc->getHeader('pages', $this->pageinfo, $this->pageinfo['_thePath']) . '<br />' . $LANG->sL('LLL:EXT:lang/locallang_core.xml:labels.path') . ': ' . $pathCropped;

			$this->content.=$this->doc->startPage($LANG->getLL('title'));
			$this->content.=$this->doc->header($LANG->getLL('title'));
			$this->content.=$this->doc->spacer(5);
			$this->content.=$this->doc->section('', $this->doc->funcMenu($headerSection, t3lib_BEfunc::getFuncMenu($this->id, 'SET[function]', $this->MOD_SETTINGS['function'], $this->MOD_MENU['function'])));
			$this->content.=$this->doc->divider(5);


			// Render content:
			$this->moduleContent();


			// ShortCut
			if ($BE_USER->mayMakeShortcut()) {
				$this->content.=$this->doc->spacer(20) . $this->doc->section('', $this->doc->makeShortcutIcon('id', implode(',', array_keys($this->MOD_MENU)), $this->MCONF['name']));
			}

			$this->content.=$this->doc->spacer(10);
		} else {
			// If no access or if ID == zero

			$this->doc = t3lib_div::makeInstance('mediumDoc');
			$this->doc->backPath = $BACK_PATH;

			$this->content.=$this->doc->startPage($LANG->getLL('title'));
			$this->content.=$this->doc->header($LANG->getLL('title'));
			$this->content.=$this->doc->spacer(5);
			$this->content.=$this->doc->spacer(10);
		}
	}

	/**
	 * Prints out the module HTML
	 *
	 * @return	void
	 */
	function printContent() {

		$this->content.=$this->doc->endPage();
		echo $this->content;
	}

	function saveSubmit() {
		$priority = t3lib_div::_GP('priority');
		$period = t3lib_div::_GP('period');
		$site = t3lib_div::_GP('site');


		foreach ($priority as $uid => $val) {
			$deleted = ($site[$uid] == "on" ? 1 : 0);

			$where = "pid = " . intval($uid);
			$set = array("disabled" => $deleted, "priority" => $val, "changefreq" => $period[$uid]);
			if ($this->data[$uid]) {
				$this->db->exec_UPDATEquery("tx_weeaargooglesitemap", $where, $set);
			} else {
				$set["pid"] = $uid;
				$res = $this->db->exec_INSERTquery("tx_weeaargooglesitemap", $set);
			}
		}
	}

	function getData() {
		$res = $this->db->exec_SELECTquery("*", "tx_weeaargooglesitemap", "1=1");
		while ($row = $GLOBALS['TYPO3_DB']->sql_fetch_assoc($res)) {
			$this->data[$row['pid']] = array($row['disabled'], $row['priority'], $row['changefreq']);
		}
	}

	/**
	 * Generates the module content
	 *
	 * @return	void
	 */
	function moduleContent() {
		$this->db = $GLOBALS["TYPO3_DB"];

		switch ((string) $this->MOD_SETTINGS['function']) {
			case 1:
				$this->getData();

				if (t3lib_div::_GP('priority')) {
					$this->saveSubmit();
				}
				$this->getData();

				$this->showPageTree();
				break;
			case 2:
				$content = '<div align=center><strong>Menu item #2...</strong></div>';
				$this->content.=$this->doc->section('Message #2:', $content, 0, 1);
				break;
			case 3:
				$content = '<div align=center><strong>Menu item #3...</strong></div>';
				$this->content.=$this->doc->section('Message #3:', $content, 0, 1);
				break;
		}
	}

	function showPageTree() {
		$siteRootUid = $this->id;
		if (isset($fromRoot))
			$siteRootUid = $this->getSiteRoot();

		if ($siteRootUid) {
			$pages = $this->getPageTreeOrder($siteRootUid);

			$this->content .= $this->getPageInfo($pages);
		}
	}

	function getPriority($uid, $wrap = "<option %s>|</option>") {
		$content = "";
		for ($i = 0; $i < 1; $i+=0.1) {
			if (strlen($this->data[$uid][1]) > 0) {

				// Workaround, because int compare doesn't work
				if (str_replace(".", ",", $this->data[$uid][1]) == str_replace(".", ",", $i)) {
					$select = "selected";
				} else {
					$select = "";
				}
			} else {
				$select = ($i == 0.5 ? "selected" : "");
			}

			$content .= str_replace("|", $i, sprintf($wrap, $select));
		}

		return $content;
	}

	function getPeriod($uid, $wrap = "<option %s>|</option>") {
		$period = array('', 'always', 'hourly', 'daily', 'weekly', 'monthly', 'yearly', 'never');

		foreach ($period as $val) {
			$select = ($this->data[$uid][2] == $val ? "selected" : "");
			$content .= str_replace("|", $val, sprintf($wrap, $select));
		}

		return $content;
	}

	function getPageInfo($pages, $level = 0) {
		$allowed_doktype = array(2, 1, 4);

		$content = "<form action='' method='post'>";
		$content .= "<table>";
		$content .= "<tr><td><strong>Disabled</strong></td><td><strong>Doktype</strong></td><td><strong>Page ID</strong></td><td><strong>Title</strong></td><td><strong>Priority</strong></td><td><strong>Period</strong><td></tr>";
		for ($i = 1; $i < count($pages); $i++) {
			$select = ($this->data[$pages[$i]['uid']][0] == 1 ? "checked" : "");

			$content .= "<tr>";
			if (in_array($pages[$i]['doktype'], $allowed_doktype)) {
				$content .= "<td><input type='checkbox' name='site[" . $pages[$i]['uid'] . "]' " . $select . "/></td>";
			} else {
				$content .= "<td>&nbsp;</td>";
			}
			$content .= "<td>" . $pages[$i]['doktype'] . "</td>";
			$content .= "<td>" . $pages[$i]['uid'] . "</td>";
			if (!isset($startLevel)) {
				$startLevel = $pages[$i]['level'];
			}

			$content .= "<td>";

			if ($pages[($i + 1)]['level'] != $pages[$i]['level'] && $pages[($i + 1)]['level'] >= $startLevel) {
				for ($j = 2; $j < $pages[$i]['level']; $j++) {
					$content .= "<img src='img/sub1.gif'>";
				}
				$content .= "<img src='img/sub.gif'>";
			} else {
				for ($j = 1; $j < $pages[$i]['level']; $j++) {
					$content .= "<img src='img/sub1.gif'>";
				}
				$content .= "<img src='img/sub.gif'>";
			}
			$content .= $pages[$i]['title'];
			$content .= "</td>";
			if (in_array($pages[$i]['doktype'], $allowed_doktype)) {
				$content .= "<td><select name=\"priority[" . $pages[$i]['uid'] . "]\">" . $this->getPriority($pages[$i]['uid']) . "</select></td>";
				$content .= "<td><select name=\"period[" . $pages[$i]['uid'] . "]\">" . $this->getPeriod($pages[$i]['uid']) . "</select></td>";
			} else {
				$content .= "<td>&nbsp;</td><td>&nbsp;</td>";
			}
			$content .= "</tr>";
		}
		$content .= "<tr><td colspan='4'><input type='submit'/></td></tr>";
		$content .= "</table>";
		$content .= "</form>";


		return $content;
	}

	function getPagetreeOrder($siteRootUid) {
		$pages[] = array();
		$query = "deleted='0' AND hidden='0' ORDER BY sorting,pid";
		$res = $GLOBALS['TYPO3_DB']->exec_SELECTquery('uid, pid, sorting, title,doktype', 'pages', $query);
		while ($row = $GLOBALS['TYPO3_DB']->sql_fetch_assoc($res)) {
			$tmp = array(
				'uid' => $row['uid'],
				'pid' => $row['pid'],
				'title' => $row['title'],
				'doktype' => $row['doktype'],
				'additional' => '0'
			);
			$pages[] = $tmp;
		}
		$startPid = $siteRootUid;
		$orderNum = -1;
		$order = array(array('uid' => $siteRootUid, 'level' => '0', 'title' => '', 'doktype' => ''));
		// l0:
		$l0 = 0;
		while ($l0 < count($pages)) {
			if ($pages[$l0]['uid'] == $startPid) {
				$curL0Uid = $pages[$l0]['uid'];
				$order[] = array('uid' => $curL0Uid, 'level' => '1', 'title' => $pages[$l0]['title'], 'doktype' => $pages[$l0]['doktype']);
				//l1:
				$l1 = 0;
				while ($l1 < count($pages)) {
					if ($pages[$l1]['pid'] == $curL0Uid) {
						$curL1Uid = $pages[$l1]['uid'];
						$order[] = array('uid' => $curL1Uid, 'level' => '2', 'title' => $pages[$l1]['title'], 'doktype' => $pages[$l1]['doktype']);
						//l2:
						$l2 = 0;
						while ($l2 < count($pages)) {
							if ($pages[$l2]['pid'] == $curL1Uid) {
								$curL2Uid = $pages[$l2]['uid'];
								$order[] = array('uid' => $curL2Uid, 'level' => '3', 'title' => $pages[$l2]['title'], 'doktype' => $pages[$l2]['doktype']);
								//l3:
								$l3 = 0;
								while ($l3 < count($pages)) {
									if ($pages[$l3]['pid'] == $curL2Uid) {
										$curL3Uid = $pages[$l3]['uid'];
										$order[] = array('uid' => $curL3Uid, 'level' => '4', 'title' => $pages[$l3]['title'], 'doktype' => $pages[$l3]['doktype']);
										//l4:
										$l4 = 0;
										while ($l4 < count($pages)) {
											if ($pages[$l4]['pid'] == $curL3Uid) {
												$curL4Uid = $pages[$l4]['uid'];
												$order[] = array('uid' => $curL4Uid, 'level' => '5', 'title' => $pages[$l4]['title'], 'doktype' => $pages[$l4]['doktype']);
												//l5:
												$l5 = 0;
												while ($l5 < count($pages)) {
													if ($pages[$l5]['pid'] == $curL4Uid) {
														$curL5Uid = $pages[$l5]['uid'];
														$order[] = array('uid' => $curL5Uid, 'level' => '6', 'title' => $pages[$l5]['title'], 'doktype' => $pages[$l5]['doktype']);
													}
													$l5++;
												} // end l5
											}
											$l4++;
										}   //end l4
									}
									$l3++;
								}   //end l3
							}
							$l2++;
						}   //end l2
					}
					$l1++;
				} //end l1
			}
			$l0++;
		} //end l0
		return $order;
	}

}

if (defined('TYPO3_MODE') && $TYPO3_CONF_VARS[TYPO3_MODE]['XCLASS']['ext/weeaar_googlesitemap/mod1/index.php']) {
	include_once($TYPO3_CONF_VARS[TYPO3_MODE]['XCLASS']['ext/weeaar_googlesitemap/mod1/index.php']);
}




// Make instance:
$SOBE = t3lib_div::makeInstance('tx_weeaargooglesitemap_module1');
$SOBE->init();

// Include files?
foreach ($SOBE->include_once as $INC_FILE)
	include_once($INC_FILE);

$SOBE->main();
$SOBE->printContent();
?>


ext_emconf.php
<?php

########################################################################
# Extension Manager/Repository config file for ext: "weeaar_googlesitemap"
#
# Auto generated 02-05-2008 13:47
#
# Manual updates:
# Only the data in the array - anything else is removed by next write.
# "version" and "dependencies" must not be touched!
########################################################################

$EM_CONF[$_EXTKEY] = array(
	'title' => 'Google Sitemap',
	'description' => 'create a xml sitemap for google',
	'category' => 'plugin',
	'shy' => 0,
	'version' => '0.4.4',
	'dependencies' => 'cms',
	'conflicts' => '',
	'priority' => '',
	'loadOrder' => '',
	'module' => 'mod1',
	'state' => 'beta',
	'uploadfolder' => 0,
	'createDirs' => '',
	'modify_tables' => '',
	'clearcacheonload' => 0,
	'lockType' => '',
	'author' => 'Marcel Alburg',
	'author_email' => 'alb@weeaar.com',
	'author_company' => '',
	'CGLcompliance' => '',
	'CGLcompliance_note' => '',
	'constraints' => array(
		'depends' => array(
			'cms' => '',
			'typo3' => '4.5.0-6.2.99',
		),
		'conflicts' => array(
		),
		'suggests' => array(
		),
	),
	'_md5_values_when_last_written' => 'a:36:{s:9:"ChangeLog";s:4:"4e44";s:10:"README.txt";s:4:"6b44";s:12:"ext_icon.gif";s:4:"f533";s:17:"ext_localconf.php";s:4:"c8d2";s:14:"ext_tables.php";s:4:"a95b";s:14:"ext_tables.sql";s:4:"5d73";s:14:"doc/manual.sxw";s:4:"1b38";s:19:"doc/wizard_form.dat";s:4:"1775";s:20:"doc/wizard_form.html";s:4:"67f9";s:40:"pi1/class.tx_weeaargooglesitemap_pi1.php";s:4:"8790";s:14:"mod1/clear.gif";s:4:"cc11";s:13:"mod1/conf.php";s:4:"2734";s:14:"mod1/index.php";s:4:"c97d";s:18:"mod1/locallang.xml";s:4:"4557";s:22:"mod1/locallang_mod.xml";s:4:"ad68";s:19:"mod1/moduleicon.gif";s:4:"f533";s:28:"mod1/img/cache_all-trans.gif";s:4:"7d2f";s:22:"mod1/img/cache_all.gif";s:4:"ba76";s:18:"mod1/img/clear.gif";s:4:"cc11";s:21:"mod1/img/doktype1.gif";s:4:"fdf4";s:23:"mod1/img/doktype199.gif";s:4:"1b74";s:23:"mod1/img/doktype254.gif";s:4:"1aed";s:23:"mod1/img/doktype255.gif";s:4:"cf3b";s:21:"mod1/img/doktype3.gif";s:4:"a5bb";s:21:"mod1/img/doktype4.gif";s:4:"fdc3";s:21:"mod1/img/doktype5.gif";s:4:"fe6e";s:21:"mod1/img/doktype6.gif";s:4:"5fc2";s:21:"mod1/img/doktype7.gif";s:4:"9f7a";s:23:"mod1/img/helpbubble.gif";s:4:"7877";s:23:"mod1/img/infobubble.gif";s:4:"5be3";s:20:"mod1/img/refresh.gif";s:4:"ad8e";s:16:"mod1/img/sub.gif";s:4:"099a";s:17:"mod1/img/sub1.gif";s:4:"2d14";s:19:"mod1/img/sub1_1.gif";s:4:"8c2f";s:19:"mod1/img/sub1_2.gif";s:4:"c2b7";s:16:"mod1/img/top.gif";s:4:"6360";}',
);

?>